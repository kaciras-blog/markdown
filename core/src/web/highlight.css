/* 跟我的 IDE 差不多的配色，为了醒目加亮了点。 */

.hljs {
	--color: #ebf2f5;
	background: #2b2b2b;
	color: var(--color);
}

.hljs-insert {
	background: rgba(0, 255, 128, .15);
	padding-left: 6px;

	&::before {
		content: "+";
		color: lightgreen;
	}
}

.hljs-delete {
	background: rgba(255, 50, 50, .25);
	padding-left: 6px;

	&::before {
		content: "-";
		color: red;
	}
}

.hljs-title,
.hljs-number,
.hljs-tag,
.hljs-subst,
.hljs-selector-pseudo,
.hljs-selector-class {
	color: var(--color);
}

.hljs-class .hljs-title {
	color: var(--color);
}

.hljs-deletion,
.hljs-strong,
.hljs-emphasis {
	color: #a8a8a2;
}

.hljs-bullet,
.hljs-quote,
.hljs-regexp,
.hljs-literal,
.hljs-link {
	color: #ae81ff;
}

.hljs-code,
.hljs-section {
	color: #a6e22e;
}

.hljs-strong {
	font-weight: bold;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-keyword,
.hljs-type,
.hljs-selector-tag,
.hljs-built_in,
.hljs-builtin-name {
	color: #f558e8;
}

.hljs-name{
	color: #E6AE6E;
}

.hljs-symbol,
.hljs-attribute,
.hljs-attr,
.hljs-meta {
	color: #67e9f0;
}

.hljs-variable,
.hljs-params {
	color: #fff394;
}

.hljs-string,
.hljs-selector-id,
.hljs-selector-attr,
.hljs-addition,
.hljs-template-variable {
	color: #fa648c;
}

.hljs-comment {
	color: #75eb75
}
