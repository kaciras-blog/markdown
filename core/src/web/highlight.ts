import hljs from "highlight.js/lib/core";

// 只添加我常用的语言。
import c from "highlight.js/lib/languages/c";
import cpp from "highlight.js/lib/languages/cpp";
import xml from "highlight.js/lib/languages/xml";
import cs from "highlight.js/lib/languages/csharp";
import css from "highlight.js/lib/languages/css";
import dockerfile from "highlight.js/lib/languages/dockerfile";
import go from "highlight.js/lib/languages/go";
import http from "highlight.js/lib/languages/http";
import ini from "highlight.js/lib/languages/ini";
import java from "highlight.js/lib/languages/java";
import javascript from "highlight.js/lib/languages/javascript";
import json from "highlight.js/lib/languages/json";
import kotlin from "highlight.js/lib/languages/kotlin";
import less from "highlight.js/lib/languages/less";
import lua from "highlight.js/lib/languages/lua";
import protobuf from "highlight.js/lib/languages/protobuf";
import python from "highlight.js/lib/languages/python";
import rust from "highlight.js/lib/languages/rust";
import scss from "highlight.js/lib/languages/scss";
import shell from "highlight.js/lib/languages/shell";
import sql from "highlight.js/lib/languages/sql";
import yaml from "highlight.js/lib/languages/yaml";
import typescript from "highlight.js/lib/languages/typescript";
import vue from "./hljs-vue.ts";

hljs.registerLanguage("c", c);
hljs.registerLanguage("cpp", cpp);
hljs.registerLanguage("xml", xml);
hljs.registerLanguage("cs", cs);
hljs.registerLanguage("css", css);
hljs.registerLanguage("dockerfile", dockerfile);
hljs.registerLanguage("go", go);
hljs.registerLanguage("http", http);
hljs.registerLanguage("ini", ini);
hljs.registerLanguage("java", java);
hljs.registerLanguage("javascript", javascript);
hljs.registerLanguage("json", json);
hljs.registerLanguage("kotlin", kotlin);
hljs.registerLanguage("less", less);
hljs.registerLanguage("lua", lua);
hljs.registerLanguage("protobuf", protobuf);
hljs.registerLanguage("python", python);
hljs.registerLanguage("rust", rust);
hljs.registerLanguage("scss", scss);
hljs.registerLanguage("shell", shell);
hljs.registerLanguage("sql", sql);
hljs.registerLanguage("yaml", yaml);
hljs.registerLanguage("typescript", typescript);
hljs.registerLanguage("vue", vue);

/*
 * # 差分的实现
 * Prism.js 虽有个插件，但它使用 Git 的补丁语法，跟我需要的不一样。
 * 另外 Prism 对 TypeScript 的识别不佳，故仍然用 Highlight.js。
 * https://prismjs.com/plugins/diff-highlight
 *
 * 由于 Highlight.js 没有原生和插件支持，作者也没实现的想法，所以只能自己做了。
 * https://github.com/highlightjs/highlight.js/issues/480
 */

export default function (code: string, language: string, attrs?: string) {
	if (!hljs.getLanguage(language)) {
		return undefined;
	}
	if (attrs !== "diff") {
		return hljs.highlight(code, { language }).value;
	}

	// 遍历每一行，记录以差分符号（+/-）开头的行号。
	let lines = code.split("\n");
	const changes = new Map<number, boolean>();

	for (let i = 0; i < lines.length; i++) {
		switch (lines[i].charCodeAt(0)) {
			case 43: /* + */
				changes.set(i, true);
				break;
			case 45: /* - */
				changes.set(i, false);
				break;
			default:
				continue;
		}
		// 去掉差分符号使 highlight.js 能够处理。
		lines[i] = lines[i].slice(1);
	}

	code = lines.join("\n");
	code = hljs.highlight(code, { language }).value;
	lines = code.split("\n");

	// 底层的高亮完成后重新按行分割，并给记录的行加上标签。
	const htmlFragments = [];
	for (let i = 0; i < lines.length; i++) {
		if (changes.has(i - 1)) {
			htmlFragments.push("</span>");
		}
		switch (changes.get(i)) {
			case true:
				htmlFragments.push("<span class='hljs-insert'>");
				break;
			case false:
				htmlFragments.push("<span class='hljs-delete'>");
				break;
		}
		htmlFragments.push(lines[i], "\n");
	}

	htmlFragments.pop();
	return htmlFragments.join(""); // 最后合并行，相当于两次高亮处理。
}
