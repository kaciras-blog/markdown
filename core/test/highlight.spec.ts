import { expect, it } from "vitest";
import highlight from "../src/web/highlight.ts";

it("should support diff syntax", () => {
	expect(highlight("+A\n-B\nC", "ini", "diff")).toMatchSnapshot();
});

it("should work without diff", () => {
	expect(highlight("+A\n-B\nC", "ini")).toMatchSnapshot();
});

it("should only escape HTML for unknown language", () => {
	expect(highlight("<foo>", "UNKNOWN")).toBe("&lt;foo&gt;");
});

it("should only escape HTML without language", () => {
	expect(highlight("<foo>", "")).toBe("&lt;foo&gt;");
});
