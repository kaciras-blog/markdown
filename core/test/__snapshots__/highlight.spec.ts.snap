// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should support diff syntax 1`] = `
"<span class='hljs-insert'>A
</span><span class='hljs-delete'>B
</span>C"
`;

exports[`should work without diff 1`] = `
"+A
-B
C"
`;
